students/NatalieRodriguez/Lesson05/mailroom3.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+
+# Lesson 05: Mailroom: Part 3, With Exception Handling
+# Natalie Rodriguez
+# April 17, 2018
+
+# Requirements:
+# 1)use dictionaries, when possible
+# 2)use a dictionary to switch between user selections
+# 3)Try to use a dict and the .format() method to do the letter as one big template
+# rather than building strings in parts.
+# 4) Add file writing functionality, i.e. a function (and a menu item to invoke it)
+# to go through all the donors in your donor data structure, generate a thank you
+# and write it to disk as a text file.
+
+import os
+import pathlib
+
+# Initial donor list and the amounts they have donated
+donors = {'Luke Rodriguez':[12.75,50.31,42.59],'Emma Burgess':[390.87,30.00],'Greg Cayetano':[875.60,1120.00],
+'Hannah Watson':[20.58,1120.14],'Emily Connor':[10.00],'Catherine Davis':[400.00],
+'River Tails':[63.56,1200.00,300.65],'Virginia Ferdinand':[350.35,5000.00],'Joseph Kibson':[3498.00,5.50]}
+
+
+def donor_dashboard():
+    # dict of menu options
+    options = {'1': ("Send a thank you note.", thank_you),
+               '2': ("Create a report.", create_report),
+               '3': ("Send thank yous to all donors.", thank_all),
+               '4': ("Exit the donor dashboard.", exit)}
+
+    response = ''
+    while response != '4':
+        print()
+        for i in options:
+            print(i, options[i][0])
+
+        # Get user's selection
+        response = ''
+        while not response in options:
+            response = input("Type your selection: ").strip()
+
+        options[response][1]()  # Call helper function
+
+
+def exit():
+
+    print("\nExiting the donor dashboard. Goodbye! \n")
+    return
+
+
+def thank_you():
+    """
+    Add new donations for new or existing donors, and send a thank-you
+    letter.
+    :return:  None.
+    """
+    # Get the donor name, show all donors, or quit
+    response = input(
+        "\nType the full donor name (Enter 'list' to show all donors or enter 'exit'): "
+    ).strip()
+
+    if response.lower() in ('', 'exit'):
+        exit()
+        return
+
+    elif response.lower() == 'list':
+        print("\nDONORS:")
+        for donor in donors:
+            print(donor)
+        thank_you()
+
+    else:
+        while True:
+            donation = input(
+                f"Enter amount of donation by '{response}' (or enter 'exit'): "
+            ).strip().lower()
+            if donation == 'exit':
+                exit()
+                return
+
+            elif donation.strip('0123456789.') == '' and len(
+                    donation) > 0 and donation.count('.') <= 1 and float(
+                donation) > 0.0:
+                break
+
+        donation = float(donation)
```

Comment:
This would be another good opportunity for a try/except block on the float conversion, to validate that you're getting numeric input